src/main/java/net/minecraftforge/common/smelting/SmeltingRecipeRegistry.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.smelting;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.crafting.FurnaceRecipes;
+import net.minecraftforge.oredict.OreDictionary;
+
+import com.google.common.reflect.Reflection;
+
+/**
+ * Recipe for smelting recipes, should be used over the GameRegistry methods and vanilla's FurnaceRecipes class.
+ */
+public class SmeltingRecipeRegistry
+{
+
+    /**
+     * Default time a recipe takes to cook.
+     */
+    public static final int DEFAULT_COOK_TIME = 200;
+
+    private static final List<SmeltingRecipe> recipes = new ArrayList<SmeltingRecipe>();
+    private static final List<SmeltingRecipe> recipesUn = Collections.unmodifiableList(recipes);
+
+    static
+    {
+        // ensure vanilla recipes are first
+        Reflection.initialize(FurnaceRecipes.class);
+    }
+
+    /**
+     * Get an unmodifiable list of all smelting recipes. The list will update in-place as new recipes are added.
+     *
+     * @return the list of recipes
+     */
+    public static List<SmeltingRecipe> getRecipes()
+    {
+        return recipesUn;
+    }
+
+    /**
+     * Get the first recipe matching the given input stack or null if no recipe matches.
+     *
+     * @param input the input stack
+     * @return a matching recipe
+     */
+    @Nullable
+    public static SmeltingRecipe getMatchingRecipe(@Nonnull ItemStack input)
+    {
+        for (SmeltingRecipe recipe : recipes)
+        {
+            if (recipe.matches(input))
+            {
+                return recipe;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Add the given recipe.
+     *
+     * @param recipe the recipe
+     */
+    public static void addRecipe(@Nonnull SmeltingRecipe recipe)
```

Comment:
All these `addRecipe` methods should support an XP parameter that gets used somehow, otherwise these are not a direct replacement of the `GameResgitry.addSmelting` methods because there is no support for recipes with XP.